File: binance4j-core/src/test/java/com/binance4j/core/event/IntervalEventTest.java
package com.binance4j.core.event;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.time.Duration;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import com.binance4j.core.test.ConcurrentTest;

class IntervalEventTest extends ConcurrentTest {
    CompletableFuture<Integer> future;
    boolean completed;
    ScheduledEvent event;
    final int maxTicks = 2;

    @BeforeEach
    public void initFuture() {
        future = new CompletableFuture<>();
    }

    @Test
    @DisplayName("The event should run the given number of times")
    void testFixedExecution() throws InterruptedException, ExecutionException {

        // completes after the event is supposed to be called
        new Timer().schedule(new CompleteTask(), Duration.ofSeconds(3).toMillis());

        event = new IntervalEvent(Duration.ofSeconds(1), maxTicks, () -> System.out.printf("TimeoutEvent's task called %s times%n", event.getTicks()));

        assertEquals(maxTicks, future.get());
    }

    @Test
    @DisplayName("The event should run indefinitely")
    void testIndefinitelyExecution() throws InterruptedException, ExecutionException {
        event = new IntervalEvent(Duration.ofSeconds(1), () -> System.out.printf("TimeoutEvent's task called %s times%n", event.getTicks()));

        // completes after the event is supposed to be called
        new Timer().schedule(new CompleteTask(), Duration.ofSeconds(6).toMillis());

        assertNotNull(future.get());
    }

    class CompleteTask extends TimerTask {
        @Override
        public void run() {

            future.complete(event.getTicks());
        }
    }
}

File: binance4j-core/src/test/kotlin/com/binance4j/core/event/IntervalEventTest.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.core.event

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.time.Duration
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException

internal class IntervalEventTest {
    private val maxTicks: Int = 2

    @Test
    @Throws(InterruptedException::class, ExecutionException::class)
    fun testFixedExecution() {
        val future = CompletableFuture<Int>()
        val task = MyTask()
        val event = IntervalEvent(task, Duration.ofSeconds(1), maxTicks)

        task.event = event

        Timer().schedule(CompleteTask(future, event), Duration.ofSeconds(3).toMillis())
        Assertions.assertEquals(maxTicks, future.get())
    }

    @Test
    @Throws(InterruptedException::class, ExecutionException::class)
    fun testIndefinitelyExecution() {
        val future = CompletableFuture<Int>()
        val task = MyTask()
        val event = IntervalEvent(task, Duration.ofSeconds(1))

        task.event = event

        Timer().schedule(CompleteTask(future, event), Duration.ofSeconds(6).toMillis())
        Assertions.assertNotNull(future.get())
    }

    internal inner class CompleteTask(val future: CompletableFuture<Int>, var event: IntervalEvent) : TimerTask() {
        override fun run() {
            future.complete(event.ticks)
        }
    }

    internal inner class MyTask : ScheduledTask {
        lateinit var event: IntervalEvent

        override fun call() {
            if (::event.isInitialized) {
                System.out.printf("TimeoutEvent's task called %s times%n", event.ticks)
            }
        }

    }
}